README
NAME
    Object::Deadly - An object that dies whenever examined

VERSION
    Version 0.05

SYNOPSIS
    This object is meant to be used in testing. All possible overloading and
    method calls die. You can pass this object into methods which are not
    supposed to accidentally trigger any potentially overloading.

      use Object::Deadly;
      
  my $foo = Object::Deadly->new;
      print $foo; # dies

METHODS
    "Object::Deadly->new()"
    "Object::Deadly->new( MESSAGE )"
        The class method "Object::Deadly->new" returns an "Object::Deadly"
        object. Dies with a stack trace and a message when evaluated in any
        context. The default message contains a stack trace from where the
        object is created.

    "Object::Deadly->kill_function( FUNCTION NAME )"
    "Object::Deadly->kill_function( FUNCTION NAME, DEATH CODE REF )"
        The class method kill_function accepts a function name like "isa",
        "can", or similar and creates a function in the
        "Object::Deadly::_unsafe" class of the same name.

        An optional second argument is a code reference to die with. This
        defaults to "\&Object::Deadly::_death".

    "Object::Deadly->kill_UNIVERSAL"
        This class method kills all currently known UNIVERSAL functions so
        they can't be called on a "Object::Deadly" object.

    "Object::Deadly->get_death"
        Returns the function "Object::Deadly::_death".

PRIVATE FUNCTIONS
    The following functions are all private and not meant for public
    consumption.

    "_death( $obj )"
        This function temporarilly reblesses the object into
        "Object::Deadly::_safe", extracts the message from inside of it, and
        "confess"'s with it. If possible this will be Carp::Clan::confess.

AUTHOR
    Joshua ben Jore, "<jjore at cpan.org>"

BUGS
    Please report any bugs or feature requests to "bug-object-deadly at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Object-Deadly>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

      perldoc Object::Deadly

    You can also look for information at:

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Object-Deadly>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Object-Deadly>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Object-Deadly>

    *   Search CPAN

        <http://search.cpan.org/dist/Object-Deadly>

ACKNOWLEDGEMENTS
COPYRIGHT & LICENSE
    Copyright 2006 Joshua ben Jore, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Object-Deadly

0.01    2006-08-14
        First version, released on an unsuspecting world.

0.02    2006-08-14
        I hate svk.

0.03    2006-08-15
        Use Devel::StackTrace for better stack traces by default.

0.04    2006-08-15
        C<...> -> C<< ... >> fixed pod formatting
        Use Carp::Clan 5.4+ or failover to Carp.
        Expanded author tests.

0.05    2006-08-15
        Created ::_safe and ::_unsafe packages for the implementation.
        Allowed Object::Deadly to get subclassed.
        Added a list of all known UNIVERSAL methods
        Added API functions ->get_death, ->kill_UNIVERSAL, ->kill_function

0.06    2006-08-15
        I'm done now? Right?
        Removed the version from the pod.
        Moved the pod to the end. It was distracting.
        Made t/20deadly.y actually run. It wasn't before.

0.07    2006-09-12
        Added some trivial documentation to Object::Deadly::_safe.
        Made Object::Deadly::_safe ok for Lint's StrictOO check.
        Removed all the isa cache stuff til I can do that sanely.
        Added a bit more docs.
        Added acknowledgement section.


Makefile.PL
use strict;
use ExtUtils::MakeMaker;

my %prerequisites;

print <<"END_NOTE";

Carp::Clan 5.4+ allows the stack traces to ignore Object::Deadly so
you get cleaner output. The 5.4+ version might not be on CPAN but is
accessible by applying the patch at
http://rt.cpan.org//Ticket/Display.html?id=21002.

END_NOTE
if (prompt(
        'Do you wish to add Carp::Clan 5.4+ to the list of prerequisites?',
        'n' ) =~ /y/i
    )
{
    $prerequisites{'Carp::Clan'} = '5.4';
}

WriteMakefile(
    NAME          => 'Object::Deadly',
    AUTHOR        => 'Joshua ben Jore <jjore@cpan.org>',
    VERSION_FROM  => 'lib/Object/Deadly.pm',
    ABSTRACT_FROM => 'lib/Object/Deadly.pm',
    PREREQ_PM     => {
        'Test::Exception'   => 0,
        'Devel::Symdump'    => 0,
        'Devel::StackTrace' => 0,
        %prerequisites,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'Object-Deadly-*' },
);


lib/Object/Deadly/_unsafe.pm
## no critic (Version,PodSections,Warnings,Rcs)
package Object::Deadly::_unsafe;

use strict;

use overload ();
my $death = Object::Deadly->get_death;
overload->import(
    map { $_ => $death }
        map { split ' ' }    ## no critic EmptyQuotes
        values %overload::ops    ## no critic PackageVars
);

# Kill off all UNIVERSAL things and try it at several points during
# execution just in case someone added something along the way.
use Object::Deadly ();
Object::Deadly->kill_UNIVERSAL;

#CHECK { Object::Deadly->kill_UNIVERSAL; }
#INIT  { Object::Deadly->kill_UNIVERSAL; }
#END   { Object::Deadly->kill_UNIVERSAL; }

#Object::Deadly->kill_function('AUTOLOAD');

# DESTROY is the only legal method for these objects. It has to be.
sub DESTROY { }

1;

__END__

=head1 NAME

Object::Deadly::_unsafe - Implementation for the deadly object

=head1 METHODS

=over

=item C<< $obj->DESTROY >>

The DESTROY method doesn't die. This is defined so it won't be
AUTOLOADed or fetched from UNIVERSAL.

=item C<< $obj->isa >>

=item C<< $obj->can >>

=item C<< $obj->version >>

=item C<< $obj->DOES >>

=item C<< $obj->import >>

=item C<< $obj->require >>

=item C<< $obj->use >>

=item C<< $obj->blessed >>

=item C<< $obj->dump >>

=item C<< $obj->peek >>

=item C<< $obj->refaddr >>

=item C<< $obj->exports >>

=item C<< $obj->moniker >>

=item C<< $obj->plural_moniker >>

=item C<< $obj->which >>

=item C<< $obj->AUTOLOAD >>

Each of AUTOLOAD, a named list of known UNIVERSAL functions and then a
query for everything currently known are all implemented with C<<
Object::Deadly->get_death >> to prevent anything from sneaking through
to a successful call against something in UNIVERSAL.

That list of functions are what core perl uses plus a bunch from CPAN
modules including L<UNIVERSAL>, L<UNIVERSAL::require>,
L<UNIVERSAL::dump>, L<UNIVERSAL::exports>, L<UNIVERSAL::moniker>,
L<UNIVERSAL::which>. That's just the list as it exists today. If
someone else creates a new one and you load it, be sure to do it
*prior* to loading this module so I can have at least a chance at
noticing anything it's loaded.

=back

=head1 SEE ALSO

L<Object::Deadly>, L<Object::Deadly::_safe>

=cut

1;


lib/Object/Deadly.pm
## no critic (PodSections,UseWarnings,RcsKeywords)
package Object::Deadly;

use strict;

use Devel::Symdump ();

BEGIN {

    # Fetch either Carp::Clan::confess or autoload
    # Carp::confess. Carp::Clan has a bug in 5.3 and earlier that
    # prevents it from handling overloaded objects in the call
    # stack. The ticket
    # http://rt.cpan.org//Ticket/Display.html?id=21002 proposes a fix
    # to Carp::Clan.
    #
    # Also, this happens in BEGIN because I want confess() to exist by
    # the time the parser sees the call to it farther down below.

    eval 'use Carp::Clan 5.4;';    ## no critic
    if ( not defined &confess ) {
        require autouse;
        autouse->import( Carp => 'confess' );
    }
}

use vars '$VERSION';               ## no critic Interpolation
$VERSION = '0.07';

sub new {

    # Public, overridable class method. Returns an _unsafe object.

    my $class                = shift @_;
    my $implementation_class = "$class\::_unsafe";

    my $data;
    if (@_) {
        $data = shift @_;
    }
    else {

        # No sense in loading this unless we actually use it.
        require Devel::StackTrace;

        $data = Devel::StackTrace->new( ignore_package => $class )->as_string;
        $data =~ s/\AT/Object::Deadly t/xm;

    }

    return bless \$data, $implementation_class;
}

sub kill_function {

    # Public, overridable class method. Creates a deadly function in
    # the _unsafe class.

    my ( $class, $func, $death ) = @_;
    my $implementation_class = "$class\::_unsafe";
    my $function_name        = "$implementation_class\::$func";
    no strict 'refs';    ## no critic Strict

    if ( defined &$function_name ) {    ## no critic
        return;
    }
    elsif ( defined $death ) {
        *$function_name = $death;       ## no critic
        return 1;
    }
    else {
        *$function_name = $class->get_death;    ## no critic
        return 1;
    }
}

sub kill_UNIVERSAL {

    # Public, overridable method call. Creates deadly functions to
    # mask all UNIVERSAL methods.

    my $class = shift @_;
    for my $fqf_function (

        # core perl
        'isa',
        'can',
        'VERSION',

        # core perl 5.9.3+
        'DOES',

        # UNIVERSAL.pm
        'import',

        # UNIVERSAL/require.pm
        'require',
        'use',

        # UNIVERSAL/dump.pm
        'blessed',
        'dump',
        'peek',
        'refaddr',

        # UNIVERSAL/exports.pm
        'exports',

        # UNIVERSAL/moniker.pm
        'moniker',
        'plural_moniker',

        # UNIVERSAL/which.pm
        'which',

        # Anything else we happen to find
        Devel::Symdump->rnew('UNIVERSAL')->functions
        )
    {
        my $function = $fqf_function;
        $function =~ s/\AUNIVERSAL:://mx;

        $class->kill_function($function);
    }

    return 1;
}

sub get_death {

    # Public, overridable method call. Returns the _death function
    my $class = shift @_;
    no strict 'refs';    ## no critic Strict
    return \&_death;
}

sub _death {             ## no critic RequireFinalReturn
                         # The common death
    my $self = shift @_;

    # Fetch the message in the object by switching the object into
    # something that's safe.
    my $unsafe_implementation_class = ref $self;
    my $safe_implementation_class   = $unsafe_implementation_class;
    $safe_implementation_class =~ s/\::_unsafe\z/::_safe/mx;

    bless $self, $safe_implementation_class;
    my $message = $$self;    ## no critic DoubleSigils
    bless $self, $unsafe_implementation_class;

    confess "[[[[ $message ]]]]";
}

# Compile and load our implementing classes. Don't use use() because
# that'll call ->import.
use Object::Deadly::_safe   ();
use Object::Deadly::_unsafe ();

## no critic (EndWithOne)
'For the SAKE... of the FUTURE of ALL... mankind... I WILL have
a... SMALL sprite!';

__END__

=head1 NAME

Object::Deadly - An object that dies whenever examined

=head1 SYNOPSIS

  use Object::Deadly;
  use Test::Exception 'lives_ok';
  
  # Test that a few functions inspect their parameters safely
  lives_ok { some_function( Object::Deadly->new ) } 'some_function';
  lives_ok { Dumper( Object::Deadly->new ) } 'Data::Dumper';

=head1 DESCRIPTION

This object is meant to be used in testing. All possible overloading
and method calls die. You can pass this object into methods which are
not supposed to accidentally trigger any potentially overloading.

This problem arose when testing L<Data::Dump::Streamer> and
L<Carp>. The former was triggering overloaded object methods instead
of just dumping their data. L<Data::Dump::Streamer> is now safe for
overloaded objects but it wouldn't have been unless it hadn't have
been tested with a deadly, overloaded object.

=head1 DEALING WITH DEATH

TODO

=head1 METHODS

=over

=item C<< Object::Deadly->new() >>

=item C<< Object::Deadly->new( MESSAGE ) >>

The class method C<< Object::Deadly->new >> returns an C<< Object::Deadly >>
object. Dies with a stack trace and a message when evaluated in any
context. The default message contains a stack trace from where the
object is created.

=item C<< Object::Deadly->kill_function( FUNCTION NAME ) >>

=item C<< Object::Deadly->kill_function( FUNCTION NAME, DEATH CODE REF ) >>

The class method kill_function accepts a function name like C<< isa >>,
C<< can >>, or similar and creates a function in the
C<< Object::Deadly::_unsafe >> class of the same name.

An optional second argument is a code reference to die with. This
defaults to C<< Object::Deadly->can( '_death' ) >>.

=item C<< Object::Deadly->kill_UNIVERSAL >>

This class method kills all currently known UNIVERSAL functions so
they can't be called on a C<< Object::Deadly >> object. This includes
a list of methods known to the author and then an inspection of
UNIVERSAL::.

=item C<< Object::Deadly->get_death >>

Returns the function C<< Object::Deadly::_death >>.

=back

=head1 PRIVATE FUNCTIONS

The following functions are all private and not meant for public
consumption.

=over

=item C<< _death( $obj ) >>

This function temporarilly reblesses the object into
C<< Object::Deadly::_safe >>, extracts the message from inside of it,
and C<< confess >>'s with it. If possible this will be
L<Carp::Clan>::confess.

=back

=head1 AUTHOR

Joshua ben Jore, C<< <jjore at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<< bug-object-deadly at
rt.cpan.org >>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Object-Deadly>. I
will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

  perldoc Object::Deadly

You can also look for information at:

=over

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Object-Deadly>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Object-Deadly>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Object-Deadly>

=item * Search CPAN

L<http://search.cpan.org/dist/Object-Deadly>

=back

=head1 ACKNOWLEDGEMENTS

Yves Orton and Yitzchak Scott-Thoennes.

=head1 COPYRIGHT & LICENSE

Copyright 2006 Joshua ben Jore, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
